lib/Jit/LLILCJit.cpp

```diff
@@ -394,3 +395,107 @@ void __cdecl LLILCJit::fatal(int Errnum, ...) {
 void LLILCJit::signalHandler(void *Cookie) {
   // do nothing
 }
+
+template <typename ObjSetT, typename LoadResult>
+void ObjectLoadListener::operator()(llvm::orc::ObjectLinkingLayerBase::ObjSetHandleT ObjHandles,
+                const ObjSetT &Objs, const LoadResult &LoadedObjInfos) {
+  int I = 0;
+  for (const std::unique_ptr<object::ObjectFile> &PObj : Objs) {
+    const object::ObjectFile &Obj = *PObj;
+    const RuntimeDyld::LoadedObjectInfo &L = *LoadedObjInfos[I];
+
+    OwningBinary<ObjectFile> DebugObjOwner = L.getObjectForDebug(Obj);
+    const ObjectFile &DebugObj = *DebugObjOwner.getBinary();
+
+    // Get the address of the object image for use as a unique identifier
+    const void *ObjData = DebugObj.getData().data();
+    DIContext *DwarfContext = new DWARFContextInMemory(DebugObj);
+
+    // Use symbol info to iterate functions in the object.
+
+    for (symbol_iterator SI = DebugObj.symbol_begin(), E = DebugObj.symbol_end();
+         SI != E; ++SI) {
+      SymbolRef::Type SymType;
+      if (SI->getType(SymType))
+        continue;
+      if (SymType == SymbolRef::ST_Function) {
+        // Function info
+        StringRef Name;
+        uint64_t Addr;
+        if (SI->getName(Name))
+          continue;
+        if (SI->getAddress(Addr))
+          continue;
+        uint64_t Size = SI->getSize();
+
+        unsigned LastDebugOffset = -1;
+        unsigned CurrentDebugEntry = 0;
+        unsigned NumDebugRanges = 0;
+        unsigned Offset = 0;
+        ICorDebugInfo::OffsetMapping *OM;
+        // unsigned LastDebugEntry = 0;
+
+        if (DwarfContext) {
+          DILineInfoTable Lines = DwarfContext->getLineInfoForAddressRange(Addr, Size);
+
+          DILineInfoTable::iterator Begin = Lines.begin();
+          DILineInfoTable::iterator End = Lines.end();
+
+          // Count offset entries. Will skip an entry if the current IL offset matches
+          // the previous offset.
+          for (DILineInfoTable::iterator It = Begin; It != End; ++It) {
+            int LineNumber = (It->second).Line;
+
+            if (LineNumber != LastDebugOffset) {
+              NumDebugRanges++;
+              LastDebugOffset = LineNumber;
+            }
+          }
+
+          // Reset offset
+          LastDebugOffset = -1;
+
+          if (NumDebugRanges > 0) {
+
+            // Allocate OffsetMapping array
```

Comment:
How? OffsetMapping has to be allocated before we can start writing to it (and this is how it was in Ryujit when I did it originally).